prow/cmd/pipeline/admission.go

```diff
@@ -0,0 +1,172 @@
+*/
+
+package main
+
+import (
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/sirupsen/logrus"
+
+	admissionapi "k8s.io/api/admission/v1beta1"
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	prowjobv1 "k8s.io/test-infra/prow/apis/prowjobs/v1"
+
+	prowjobscheme "k8s.io/test-infra/prow/client/clientset/versioned/scheme"
+)
+
+var (
+	vscheme = runtime.NewScheme()
+	codecs  = serializer.NewCodecFactory(vscheme)
+)
+
+func init() {
+	if err := prowjobscheme.AddToScheme(vscheme); err != nil {
+		logrus.Errorf("Add prow job scheme: %v", err)
+	}
+	if err := admissionapi.AddToScheme(vscheme); err != nil {
+		logrus.Errorf("Add admission API scheme: %v", err)
+	}
+	if err := admissionregistrationv1beta1.AddToScheme(vscheme); err != nil {
+		logrus.Errorf("Add admission registration schme: %v", err)
+	}
+}
+
+// runServer starts an http server on :8443, responding to /validate requests
+func runServer(cert, privateKey string) {
+	http.HandleFunc("/validate", handle)
+	s := http.Server{
+		Addr: ":8443",
+		TLSConfig: &tls.Config{
+			ClientAuth: tls.NoClientCert,
+		},
+	}
+	logrus.Error(s.ListenAndServeTLS(cert, privateKey))
+}
+
+const contentTypeJSON = "application/json"
+
+// readRequest extracts the request from the AdmissionReview reader
+func readRequest(r io.Reader, contentType string) (*admissionapi.AdmissionRequest, error) {
+	if contentType != contentTypeJSON {
+		return nil, fmt.Errorf("Content-Type=%s, expected %s", contentType, contentTypeJSON)
+	}
+
+	if r == nil {
+		return nil, fmt.Errorf("no body")
+	}
+	body, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, fmt.Errorf("read body: %v", err)
+	}
+
+	var ar admissionapi.AdmissionReview
+	deserializer := codecs.UniversalDeserializer()
+	if _, _, err := deserializer.Decode(body, nil, &ar); err != nil {
+		return nil, fmt.Errorf("decode body: %v", err)
+	}
+	return ar.Request, nil
+}
+
+// handle reads the request and writes the response
+func handle(w http.ResponseWriter, r *http.Request) {
+	req, err := readRequest(r.Body, r.Header.Get("Content-Type"))
+	if err != nil {
+		logrus.WithError(err).Error("read")
+	}
+
+	if err := writeResponse(*req, w, onlyUpdateStatus); err != nil {
+		logrus.WithError(err).Error("write")
+	}
+}
+
+type decider func(admissionapi.AdmissionRequest) (*admissionapi.AdmissionResponse, error)
+
+// writeResponse gets the response from onlyUpdateStatus and writes it to w.
+func writeResponse(ar admissionapi.AdmissionRequest, w io.Writer, decide decider) error {
+	response, err := decide(ar)
+	if err != nil {
+		logrus.WithError(err).Error("failed decision")
+		response = &admissionapi.AdmissionResponse{
+			Result: &meta.Status{
+				Message: err.Error(),
+			},
+		}
+	}
+	var result admissionapi.AdmissionReview
```

Comment:
Can we drop the admission controller part? I think that's a rather weak part of the knative-build controller

Instead I think we should write a prowjob admission controller that validates a prowjob spec does not change. We don't need one for each type of agent.

Reply:
I removed entirely the admission controller and also the cert/key arguments from the controller command line.

Do you agree to implement the generic prowjob admission controller in a separate PR?

Reply:
I think this is the goal of this issue https://github.com/kubernetes/test-infra/issues/8468.